Modifying "Tax" task to remove Single Responsibility Principle violations

ShahdLubb opened the pull request.

ShahdLubb committed:
Modifying "Tax" task to remove Single Responsibility Principle violations.

ShahdLubb committed:
Adding a class called TaxServices

1-this class is responsible of creating a TaxCalculator of the required type and apply this Tax to any product, also it is responsible of setting the flat rate tax ratio
2- only one method now is used for calculating the taxAmount using the default value for all products which could be changed using TaxServices class, I think the code needs more refactoring because the TaxPercentageDefault in FlatRateTaxCalculator could be also changed directly without using TaxService class.

ShahdLubb committed:
change flate rate tax functionality was removed from TaxServices class to FlatRateTaxCalculator

ShahdLubb committed:
Singelton principle was applied to FlatRateTaxCalculator class

since we only need one instance of this class, also this tax was applied to Product class in the constructor since tax is mandotery

WadeeSami commented on TaxServices.cs:
```diff
@@ -0,0 +1,25 @@
+        }
+        public static void ApplyTax(Product product, ITaxCalculator TaxCalculator)
+        {
+            product.TaxCalculator = TaxCalculator;
```
why to do this ?
why should external class be concerned with setting the tax calculator for the product ?

ShahdLubb replied:
I thought there would be other types of Taxes in the future and it is a good idea to handle all different types of Taxes from the same class, but yeah for the current requirements, there is no need.

WadeeSami commented on FlatRateTaxCalculator.cs:
```diff
@@ -0,0 +1,45 @@
+{
+    public class FlatRateTaxCalculator : ITaxCalculator
+    {
+        private static double _TaxPercentageDefault = 20.0;
```
I suggest to move this to a separate config file where all constants/configs can be changed in one place

ShahdLubb replied:
do you mean a  simple text file or a new class for all configurations?

ShahdLubb committed:
holding  TaxCalculator field was removed from product class,
 also some classes were created as follows:
-PriceCalculatorConfigurations was created to hold all the needed configurations.
-PriceCalculator was created to handle the logic of calculating total price.
